pkg/cri/server/container_create_linux.go

```diff
@@ -166,10 +166,23 @@ func (c *criService) containerSpec(
 
 	// Apply envs from image config first, so that envs from container config
 	// can override them.
-	env := append([]string{}, imageConfig.Env...)
+	env := []string{}
 	for _, e := range config.GetEnvs() {
 		env = append(env, e.GetKey()+"="+e.GetValue())
 	}
+	for _, imageEnv := range imageConfig.Env {
+		found := false
+		imageEnvKey := strings.Split(imageEnv, "=")[0]
+		for _, userEnv := range config.GetEnvs() {
+			if imageEnvKey == userEnv.Key {
+				found = true
+				break
+			}
+		}
+		if !found {
+			env = append(env, imageEnv)
+		}
+	}
```

Comment:
If we're going to try to merge the values rather than just append, it'd be a better approach to use a `map[string]string` and then iterate through that to convert to a `[]string` (with `=` separating key from value) than to add an O(n^2) search.  (In addition to being more efficient, it'd be a more readable approach.)

Reply:
Agreed. The current implementation is simply copied from `moby`. I can refine them later.

Reply:
Please see the latest commit, thanks.